src/dmd/expressionsem.d

```diff
@@ -6767,6 +6829,9 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
     override void visit(CastExp exp)
     {
+        bool parent_shared_allowed = GlobalStateExpressionSemantic.shared_allowed;
+        scope (exit) GlobalStateExpressionSemantic.shared_allowed = parent_shared_allowed;
```

Comment:
Seems like you should be adding a SCOPE flag, instead of global state.

Reply:
Thanks! I don't think I've ever used it. How does it work?

Reply:
Add a flag to https://github.com/dlang/dmd/blob/master/src/dmd/dscope.d#L44 and check/set it from the scope parameter of the expression.

Reply:
This is more granular than a scope.

Reply:
But it is still a recursive analysis, it doesn't matter where you put it.